Frontend and backend changes to get tables from DB on admin page

caiseL opened the pull request:
### Frontend

- I created a view on admin home page to show all the tables from database calling a server endpoint. 
- Another one called table page **(/admin/tables/:table_name)**, in a future it'll show metadata of that table. 
- I'm not so sure about the naming on components and widgets and I'd to change them, but I don't know how I should name them.

![imagen](https://user-images.githubusercontent.com/75282728/134066807-ba611dcb-7fc5-4df0-be59-46202ad4a34f.png)

### Backend

- Regarding server/backend, I added models, controllers and services to get tables from a database and parse them using SQL syntax.

![imagen](https://user-images.githubusercontent.com/75282728/134067255-79537506-64df-46a0-ba89-bab5ca63ef1d.png)

AlexITC commented on admin/src/main/scala/net/wiringbits/components/pages/TablePage.scala:
```diff
@@ -0,0 +1,25 @@
+
+  val component: FunctionalComponent[Props] = FunctionalComponent[Props] { props =>
+    val params = useParams()
+    val table_name = params.asInstanceOf[js.Dynamic].table_name.toString
```
Rename `table_name` to `tableName`, Scala uses camelCase

AlexITC commented on admin/src/main/scala/net/wiringbits/components/widgets/Table.scala:
```diff
@@ -0,0 +1,73 @@
+import slinky.core.FunctionalComponent
+import slinky.core.annotations.react
+
+@react object Table {
```
`Table` is likely going to get conflicts because it is a very common name, I'd call it `ExperimentalTableWidget` instead

AlexITC commented on admin/src/main/scala/net/wiringbits/components/widgets/TableList.scala:
```diff
@@ -0,0 +1,76 @@
+import slinky.core.facade.Fragment
+import slinky.web.html.className
+
+object TableList {
```
Same way, I'd name this `ExperimentalTableListWidget`

AlexITC commented on admin/src/main/scala/net/wiringbits/components/widgets/TableList.scala:
```diff
@@ -0,0 +1,76 @@
+  val component: FunctionalComponent[Props] = FunctionalComponent[Props] { props =>
+    val classes = useStyles(())
+
+    def goTo(table_name: String): Unit = {
```
`table_name` -> `tableName`

AlexITC commented on admin/src/main/scala/net/wiringbits/components/widgets/TableList.scala:
```diff
@@ -0,0 +1,76 @@
+
+    def goTo(table_name: String): Unit = {
+      val url = s"/tables/$table_name"
+      dom.window.location.href = url
```
In react, you don't use these redirects because it reloads the whole page, instead, you use the react-router, like:

```
val history = reactRouterDom.useHistory()
history.push("/dashboard")
```

[example](https://github.com/wiringbits/scala-webapp-template/blob/master/web/src/main/scala/net/wiringbits/components/widgets/SignInForm.scala#L79)

AlexITC commented on server/src/main/scala/net/wiringbits/repositories/daos/TablesDAO.scala:
```diff
@@ -0,0 +1,19 @@
+    SQL"""
+        SELECT table_name
+          FROM information_schema.tables
+        WHERE table_schema='public'
```
Be aware that this is not generic enough, you do not need to look for a solution but we need to add a `TODO` to fix it some time

AlexITC commented on server/src/main/scala/net/wiringbits/repositories/daos/TablesDAO.scala:
```diff
@@ -0,0 +1,19 @@
+        SELECT table_name
+          FROM information_schema.tables
+        WHERE table_schema='public'
+          AND table_type='BASE TABLE';
```
It is ideal that you always use a deterministic ordering, I have a [draft](https://github.com/AlexITC/notes/blob/master/development-good-practices.md#sql) for other useful hints

AlexITC commented on server/src/main/scala/net/wiringbits/repositories/models/Table.scala:
```diff
@@ -0,0 +1,7 @@
+package net.wiringbits.repositories.models
+
+case class Table(
+    table_name: String
```
`table_name` -> `tableName`, on the other hand, we know this is a table and we could just call it `name`.

I'm thinking that it could be worth naming the class like `DatabaseTable` to make it clear.

AlexITC commented on server/src/main/scala/net/wiringbits/repositories/models/Table.scala:
```diff
@@ -0,0 +1,7 @@
+    table_name: String
+)
+
+object Table {}
```
Unnecessary companion object

AlexITC commented on server/src/main/scala/net/wiringbits/repositories/models/Table.scala:
```diff
@@ -0,0 +1,7 @@
+package net.wiringbits.repositories.models
```
Given that this is core-like functionality, we'd likely prefer a different package, do not worry about it right now, just take a note for the future

AlexITC reviewed (commented):
Nice job for an initial concept, now, we need to retrieve the metadata for each table, it could be either done in the same API or a specific API per table.

Thanks!

caiseL commented:
I made a few changes.

- Changed the name of DB models on server.
- Changed the name of components on admin frontend.
- Changed the name of methods and DAOs on server.
- Created a method that brings data and fields from a table.
- A function named loader was being used more than one time in the front end, so I created a custom component
- Now formats json of the table
- Table page now brings data of the selected table. Image:
![imagen](https://user-images.githubusercontent.com/75282728/134747061-a4684d64-407b-41ed-b8f7-0abd23a6c327.png)

AlexITC reviewed (commented):
Nice progress, let's try to get the data paginated.

Thanks.

AlexITC commented on server/src/main/scala/net/wiringbits/repositories/daos/DatabaseTablesDAO.scala:
```diff
@@ -8,54 +8,60 @@ import scala.collection.mutable.ListBuffer
 
-  def all()(implicit conn: Connection): List[DatabaseTable] = {
+  def all(schema: String = "public")(implicit conn: Connection): List[DatabaseTable] = {
     // TODO: Not generic enough
```
This comment does not apply anymore

AlexITC commented on server/src/main/scala/net/wiringbits/repositories/daos/DatabaseTablesDAO.scala:
```diff
@@ -0,0 +1,61 @@
+    val columnsMetadata = new ListBuffer[ColumnMetadata]()
+
+    val statement = conn.createStatement(ResultSet.CONCUR_READ_ONLY, ResultSet.CONCUR_READ_ONLY)
+    val resultSet = statement.executeQuery("SELECT * FROM " + tableName)
```
Now, this is worse because `"SELECT * FROM " + tableName` allows SQL injection

AlexITC commented on server/src/main/scala/controllers/AdminController.scala:
```diff
@@ -43,11 +44,12 @@ class AdminController @Inject() (
 
-  def getTableMetadata(tableName: String) = handleGET { request =>
+  def getTableMetadata(tableName: String, offset: Int, limit: Int) = handleGET { request =>
+    val pagination = Pagination(offset = offset, limit = limit)
```
This certainly needs to be validated, we don't want negative offsets/limits

AlexITC commented on server/src/main/scala/net/wiringbits/repositories/daos/DatabaseTablesDAO.scala:
```diff
@@ -19,9 +19,16 @@ object DatabaseTablesDAO {
-    val resultSet = statement.executeQuery("SELECT * FROM " + tableName)
+  def getTableMetadata(tableName: String, pagination: Pagination)(implicit conn: Connection): TableMetadata = {
+    // I still haven't found how to use preparedStatement with tableName
+    val sql = f"SELECT * FROM $tableName LIMIT ? OFFSET ?"
```
You are querying the whole table just to find the metadata, this doesn't need pagination, just put `LIMIT 0` or `LIMIT 1` if `0` doesn't work.

Also, I see you use this for retrieving the data, `OFFSET` is useless unless you define a deterministic ordering

AlexITC commented on server/src/main/scala/net/wiringbits/repositories/daos/DatabaseTablesDAO.scala:
```diff
@@ -37,7 +44,7 @@ object DatabaseTablesDAO {
     } finally {
       resultSet.close()
-      statement.close()
+      preparedStatement.close()
```
This shouldn't invoke `val tableData = getTableData(resultSet)`, instead, the repository should get the metadata, then, use that to invoke `getTableData`

AlexITC commented on server/src/main/scala/net/wiringbits/services/AdminService.scala:
```diff
@@ -53,16 +54,33 @@ class AdminService @Inject() (
+  def validateTableName(tableName: String): Future[Unit] = {
+    for {
+      tables <- databaseTablesRepository.all()
+    } yield {
```
There is a simpler way:

```
val exists = tables.exists(_.name == tableName)
if (exists) ()
else throw new ...
```

AlexITC commented on server/src/main/scala/net/wiringbits/repositories/daos/DatabaseTablesDAO.scala:
```diff
@@ -19,46 +19,48 @@ object DatabaseTablesDAO {
-    val metadata = resultSet.getMetaData
-    val numberOfColumns = metadata.getColumnCount
+
+    val sql = f"SELECT * FROM $tableName LIMIT ? OFFSET ?"
```
This misses the `ORDER BY ...` part, which defines a deterministic ordering

AlexITC commented on server/src/main/scala/net/wiringbits/modules/DataExplorerModule.scala:
```diff
@@ -0,0 +1,18 @@
+  )
+}
+
+case class DataExplorerSettings(tables: List[TableSettings])
```
These models should be in a different package, like the `config` one

AlexITC commented on server/src/test/scala/controllers/AdminControllerSpec.scala:
```diff
@@ -0,0 +1,54 @@
+  }
+
+  "GET /admin/tables/aaaaaaaaaaa" should {
+    "fail when database doesn't exists" in withApiClient { client =>
```
The database exists, the table is the one that doesn't exists

AlexITC commented on server/src/main/scala/net/wiringbits/util/models/WrappedString.scala:
```diff
@@ -0,0 +1,7 @@
+package net.wiringbits.util.models
+
+trait WrappedString extends Any {
```
I think we already have this somewhere

AlexITC commented on server/src/main/scala/net/wiringbits/repositories/daos/DatabaseTablesDAO.scala:
```diff
@@ -0,0 +1,124 @@
+
+        // This is just a workaround. I think it'll be better if I use a Option[T] syntax
+        // so I'll do it later
+        cell = if (data == null) Cell("null") else Cell(data)
```
Alternatively, `Cell(Option(data).getOrElse("null"))`

AlexITC commented on server/src/main/scala/net/wiringbits/repositories/daos/DatabaseTablesDAO.scala:
```diff
@@ -0,0 +1,124 @@
+  }
+
+  def countRecordsOnTable(tableName: String)(implicit conn: Connection): Int = {
+    val SQL = s"SELECT COUNT(*) FROM $tableName"
```
This one can be definitely be done the anorm-way

adinael commented on server/src/main/scala/net/wiringbits/repositories/daos/DatabaseTablesDAO.scala:
```diff
@@ -0,0 +1,249 @@
+    """.as(tableParser.*)
+  }
+
+  def getTablesInSettings(tableSettings: DataExplorerSettings): List[DatabaseTable] = {
```
Unnecesary function, this can be done using a map function 
`    tableSettings.tables.map(x => DatabaseTable(x.tableName))
`
